core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -97,6 +97,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         case requestId => throw new KafkaException("Unknown api code " + requestId)
       }
     } catch {
+      case fee: FatalExitError => throw fee
```

Comment:
With the changes in this PR, the fetcher thread throws an exception that eventually gets caught in `ShutdownableThread.run`, which calls `FatalExitError.shutdownSystem`. Before the `shutdownSystem` call, `run` can open the latch and then the thread is dead for all purposes and cannot block the shutdown hook from making progress. Does that make sense?